OvmfPkg/VirtioKeyboardDxe/VirtioKeyboard.c

```diff
@@ -0,0 +1,1560 @@
+      break;
+
+    case KEY_LEFT:
+      Key->ScanCode    = SCAN_LEFT;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_RIGHT:
+      Key->ScanCode    = SCAN_RIGHT;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_DOWN:
+      Key->ScanCode    = SCAN_DOWN;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    case KEY_BACKSPACE:
+      Key->ScanCode    = SCAN_NULL;
+      Key->UnicodeChar = CHAR_BACKSPACE;
+      break;
+
+    case KEY_TAB:
+      Key->ScanCode    = SCAN_NULL;
+      Key->UnicodeChar = CHAR_TAB;
+      break;
+
+    case KEY_ENTER:
+      Key->ScanCode = SCAN_NULL;
+      // Key->UnicodeChar = CHAR_LINEFEED;
+      Key->UnicodeChar = CHAR_CARRIAGE_RETURN;
+      break;
+
+    case KEY_ESC:
+      Key->ScanCode    = SCAN_ESC;
+      Key->UnicodeChar = CHAR_NULL;
+      break;
+
+    default:
+      if (Dev->ShiftActive) {
+        Key->ScanCode    = MapShift[Code];
+        Key->UnicodeChar = MapShift[Code];
+      } else {
+        Key->ScanCode    = Map[Code];
+        Key->UnicodeChar = Map[Code];
+      }
+
+      if (Dev->CtrlActive) {
+        // Convert Ctrl+[a-z] and Ctrl+[A-Z] into [1-26] ASCII table entries
+        // Key->ScanCode    = SCAN_NULL;
+        if ((Key->UnicodeChar >= L'a') && (Key->UnicodeChar <= L'z')) {
+          Key->UnicodeChar = (CHAR16)(Key->UnicodeChar - L'a' + 1);
+        } else if ((Key->UnicodeChar >= L'A') && (Key->UnicodeChar <= L'Z')) {
+          Key->UnicodeChar = (CHAR16)(Key->UnicodeChar - L'A' + 1);
+        }
+      }
+
+      break;
+  }
+}
+
+// -----------------------------------------------------------------------------
+// Main function processing virtio keyboard events
+STATIC
+VOID
+EFIAPI
+VirtioKeyboardGetDeviceData (
+  IN OUT VIRTIO_KBD_DEV  *Dev
+  )
+{
+  BOOLEAN           HasData;
+  UINT8             Data[KEYBOARD_RX_BUFSIZE + 1];
+  UINT32            DataSize;
+  VIRTIO_KBD_EVENT  Event;
+  EFI_TPL           OldTpl;
+
+  for ( ; ; ) {
+    HasData = VirtioKeyboardRingGetBuffer (Dev, 0, Data, &DataSize);
+
+    // Exit if no new data
+    if (!HasData) {
+      return;
+    }
+
+    if (DataSize < sizeof (Event)) {
+      continue;
+    }
+
+    // Clearing last character is not needed as it will be overwritten anyway
+    // Dev->LastKey.ScanCode = SCAN_NULL;
+    // Dev->LastKey.UnicodeChar = CHAR_NULL;
+
+    CopyMem (&Event, Data, sizeof (Event));
+
+    OldTpl = gBS->RaiseTPL (TPL_NOTIFY);
+    switch (Event.Type) {
+      case EV_SYN:
+        // Sync event received
+        break;
+
+      case EV_KEY:
+        // Key press event received
+        // DEBUG ((DEBUG_INFO, "%a: ---------------------- \nType: %x Code: %x Value: %x\n",
+        //              __func__, Event.Type, Event.Code, Event.Value));
+
+        if (Event.Value == KEY_PRESSED) {
+          // Check key modifiers first
+          if ((Event.Code == KEY_LEFTSHIFT) || (Event.Code == KEY_RIGHTSHIFT)) {
+            Dev->ShiftActive = TRUE;

```

Comment:
You need to track key state independently for left and right shift keys (same for ctrl and alt), otherwise the tracking is not correct in some corner cases such as pressing both shift keys, then lifting one of them.

Might be easier to just have an array and track the state for all keys, i.e. `Dev->KeyActive[Event.Code] = TRUE | FALSE`.

Reply:
Implementation has been updated with this change. Thanks for pointing this out:)